protocols/kad/src/bootstrap.rs

```diff
@@ -0,0 +1,169 @@
+use futures::FutureExt;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures_timer::Delay;
+
+use crate::QueryStats;
+
+#[derive(Debug)]
+pub(crate) struct Status {
+    interval: Option<Duration>,
+    next_periodic_bootstrap: Option<Delay>,
+    waker: Option<Waker>,
+
+    bootstrap_asap: bool,
+
+    is_bootstrapping: bool,
+    any_bootstrap_succeeded: bool,
+}
+
+impl Status {
+    pub(crate) fn new(interval: Option<Duration>) -> Self {
+        Self {
+            interval,
+            next_periodic_bootstrap: interval.map(Delay::new),
+            waker: None,
+            bootstrap_asap: false,
+            is_bootstrapping: false,
+            any_bootstrap_succeeded: false,
+        }
+    }
+
+    pub(crate) fn on_new_peer_in_routing_table(&mut self) {
+        self.bootstrap_asap = true;
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake()
+        }
+    }
+
+    pub(crate) fn on_started(&mut self) {
+        self.is_bootstrapping = true;
+    }
+
+    pub(crate) fn on_result(&mut self, stats: &QueryStats) {
+        if stats.num_successes() > 0 {
+            self.any_bootstrap_succeeded = true;
+        }
+
+        self.is_bootstrapping = false;
+    }
+
+    pub(crate) fn poll_next_bootstrap(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        if self.bootstrap_asap {
```

Comment:
It might be interesting to make sure we don't make a cascade of bootstrap requests. I imagine that at startup or after the first successful bootstrap, many users will insert peers in the routing table. If we keep the current system, don't we risk having many bootstraps running at the same time ?

For example, if an end user receives 10 new peers (from configuration, for example), he will probably insert them in the routing table in a `for` loop, resulting in almost 10 bootstrap requests being triggered at approximately the same time (it might not be exactly 10 since the `poll_next_bootstrap` is probably running in another thread).

Maybe a `self.is_bootstrapping != true` check should be done at first in `poll_next_bootstrap` to make sure there is only one bootstrap at a time. Besides, if we don't do this check, it seems that `self.is_bootstrapping` is never read.

We could also use the `self.any_bootstrap_succeeded` (that is not read also I think) to either bootstrap immediately (we don't want to throttle on the first bootstrap) or to wait a little bit to bootstrap (doing so might give the user some time to add more addresses which would be interesting do have a better bootstrap) : 
```rust
struct Status {
    ...
    throttle_timer: Option<Delay>, // 10ms maybe ?,
   - bootstrap_asap // remove this field
}

    pub(crate) fn on_new_peer_in_routing_table(&mut self) {
        self.throttle_timer = self.any_bootstrap_succeeded.and_then_some(||
Delay::new(10ms) ); // the bonus of this is that the potential previous timer is reset
// so that, if the end users update the routing table multiple times
// in a short period of time, the bootstrap will happen after he finished.

        if let Some(waker) = self.waker.take() {
            waker.wake()
        }
    }


pub(crate) fn poll_next_bootstrap(&mut self, cx: &mut Context<'_>) -> Poll<()> {
    if self.bootstrapping {
        self.waker = Some(cx.waker().clone());
        return Poll::Pending;
    }

    if let Some(throttle_timer) = self.throttle_timer {
        match throttle_timer {
            Poll::Ready(()) => self.throttle_timer = None,
            Poll::Pending => return Poll::Pending,
        };
    }

   .... // same that is is currently done
}
```

Reply:
Yes the idea was to check for `is_bootstrapping` but it seems that I forgot about that! Would you be able to add this check and add a test for it aswell?